fedbiomed/researcher/job.py

```diff
@@ -475,22 +495,107 @@ def start_nodes_training_round(self,
                 timing = m['timing']
                 timing['rtime_total'] = rtime_total
 
-                r = Responses({'success': m['success'],
-                               'msg': m['msg'],
-                               'dataset_id': m['dataset_id'],
-                               'node_id': m['node_id'],
-                               'params_path': params_path,
-                               'params': params,
-                               'optimizer_args': optimizer_args,
-                               'sample_size': m["sample_size"],
-                               'encryption_factor': encryption_factor,
-                               'timing': timing})
-
-                self._training_replies[round_].append(r)
+                response = Responses({
+                    'success': m['success'],
+                    'msg': m['msg'],
+                    'dataset_id': m['dataset_id'],
+                    'node_id': m['node_id'],
+                    'params_path': params_path,
+                    'params': params,
+                    'optimizer_args': optimizer_args,
+                    'optim_aux_var': optim_aux_var,
+                    'sample_size': m["sample_size"],
+                    'encryption_factor': encryption_factor,
+                    'timing': timing,
+                })
+                self._training_replies[round_].append(response)
 
         # return the list of nodes which answered because nodes in error have been removed
         return self._nodes
 
+    def upload_agg_optimizer_aux_var(
+        self,
+        optimizer: Optimizer,
+    ) -> Tuple[Optional[str], Dict[uuid.UUID, str]]:
+        """Upload auxiliary variables emitted by a researcher-side Optimizer.
+
+        Args:
+            optimizer: Optimizer instance, auxiliary variables of which to emit
+                and upload after structuring them into multiple files, to avoid
+                information leakage as well as file redundancy.
+
+        Returns:
+            url_shared: url of a file containing auxiliary variables shared
+                across all nodes, or None (in the absence of such information).
+            url_bynode: dict mapping urls of files containing node-specific
+                auxiliary variables to the nodes' id (a missing `nodes` key
+                indicates that this node has no such information to receive).
+        """
+        # Split the information between shared and node-wise dictionaries.
+        aux_shared, aux_bynode = self._prepare_agg_optimizer_aux_var(
+            optimizer=optimizer, nodes=list(self._nodes)
```

Comment:
I got a bit confused here with `aux_shared` and `aux_bynode`, is `aux_shared` really needed here? 
Is it possible to use only `aux_bynode`, and if a aux_var needs to be  broadcasted to all Nodes, we assign same variable to all node_id in the dictionary.
That way code will be slightly convoluted (unless there is something I havenot seen)

Reply:
The idea is that the auxiliary variables sent by the researcher may either be shared across all nodes (e.g. if we implement a plug-in to adjust the node-side learning rate based on the global (aggregated) loss, or if we share the global state $c_t$ in Scaffold), or specific to each node (e.g. if we share the local states $c_{i,t}$ or the correction terms $\delta_{i,t}$ in Scaffold).

In the latter case (node-specific), we want to upload them in distinct files, so that each node downloads only their own information (both to avoid unnecessary communication costs and to prevent privacy leakage).

In the former case (shared across nodes), my implementation isolates all of this information in a dedicated file, that is uploaded only once by the researcher, and downloaded by each and every node. As a result, we save on communication costs related to the file upload.

That being said, we could do something that is more costly, but simpler, which is to merely upload one file per node, regardless of the redundancy of information between these files.

Reply:
Yes I got your idea, I was just saying that I don't see why you have these 2 dictionaries:
because  :

let's assume that i want to share the same file across all my nodes. I upload this file on the url `https:/my/url`

then we can have a variable `aux_by_nodes` = `{'node_id-1': 'https:/my/url', 'node_id-2': 'https:/my/url', 'node_id-3': 'https:/my/url'} `  if i have 3 nodes 'node_id-1', 'node_id-2', 'node_id-3' ,  and this `aux_by_nodes` is equivalent to your `aux_shared`

but again maybe there is something I am missing here

Reply:
I think that you are missing the point that there can be both shared _and_ node-specific information _at the same time_. Hence the use of two distinct variables, because in the end a node can receive between zero and two urls of files to download.

So we can have `aux_shared = {module_a: {key: val}, module_b: {key: val}}` and `aux_nodes = {node_1: {module_c: {key: val}}, node_2: {module_c: {key: val}}}` that will result in `aux_url_shared = "url_to_shared_auxvar"` and `aux_url_bynode = {"node_1": "url_to_node1_auxvar", "node_2": "url_to_node2_auxvar"}`, and eventually node 1 will receive a list with both URLs to the shared and their own aux vars; similary for node 2; etc.

Reply:
Oh I see! But then, why could you  send up to only 2 auxiliary variables for each OptiModule, and not more? I thought you could sent as many url links (in a list for instance) in `aux_nodes ` (this is what we are currently doing in Fed-BioMed).

I do think we should do the same when dealing with declearn optimodules, in order to make things as generic as possible (if in declearn you plan to implement more OptiModule that uses more than 2 auxiliary variables or if we want to implement ours)

Reply:
> That being said, I still don't see why you could not append the shared content into the file that has the node specific content

You definitely could; but it would be less efficient in terms of communication costs, because you would send some information (the shared one) `n_nodes` times rather than once. I believe that the additional code complexity is worth it, but if you do not, I can simplify the logic at the cost of that researcher-upload overhead.

Reply:
I am glad that you are convinced!

To me it is already documented, both by the docstring and inline comments of `Job.upload_agg_optimizer_aux_var`. That function's inline comments, that explain the flow along the code, are the following four lines:
```
# Split the information between shared and node-wise dictionaries.
# Upload the shared information that all nodes will download.
# Upload the node-specific information, with node-specific urls.
# Return the urls of the uploaded files.
```

Reply:
Yes sorry if i wasnot clear enough, I mean we can specify why we are doing that -> to avoid possible overhead in the communication

Reply:
Okay, I added a note to the docstring.